### PR TITLE
Correctly handle RPTRST in SOLUTION section

### PR DESCRIPTION
**Pull request by joakim-hove**
Fix for: https://github.com/OPM/opm-common/issues/1862

Thank you to @bska for providing the test case which constitutes the bulk of this PR. With some help from Equinor I have concluded that report step 1 should *not* have a restart file.

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by bska** [comment]
> With some help from Equinor I have concluded that report step 1 should _not_ have a restart file.

In that case we should remove the [`WRITE`](https://github.com/OPM/opm-common/blob/a8d8d38fbfa0de62e3bdbbf365a555c0da76b7c7/tests/parser/RestartConfigTests.cpp#L55) annotation for report step 1.

**Comment by joakim-hove** [comment]
> In that case we should remove the WRITE annotation for report step 1.

Thanks - done.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by bska** [comment]
Hm, curious.  This appears to go through the regression tests in opm-simulators and it really shouldn't.  I would expect more restart steps output for at least [`MOD4_GRP`](https://github.com/OPM/opm-tests/blob/2b8a713b1f7e59ecb98f6700e81fed2c055d2a0b/model4/MOD4_GRP.DATA#L229-L234) as a result of this change in which case the new output and the existing reference solution would not compare equal.

**Comment by joakim-hove** [comment]
> This appears to go through the regression tests in opm-simulators and it really shouldn't.

Never look a gift horse in the mouth :-) 

I agree - will look into it.

**Comment by bska** [comment]
> > This appears to go through the regression tests in opm-simulators and it really shouldn't.
> 
> I agree - will look into it.

For what it's worth, I applied the current patch and additionally made the change below to [`EclipseIO::writeTimeStep()`](https://github.com/OPM/opm-common/blob/c84b95ed85fd571babe43879f8328a7db51e9a57/src/opm/output/eclipse/EclipseIO.cpp#L190). I then got the output in [`write.restart.txt`](https://github.com/OPM/opm-common/files/5081304/write.restart.txt) from
```
grep Restart ./MOD4_GRP.PRT >write.restart.txt
```
which shows that [`getWriteRestartFile()`](https://github.com/OPM/opm-common/blob/c84b95ed85fd571babe43879f8328a7db51e9a57/opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.hpp#L340) returns `false` for all report steps `> 0` even with the current patch in this PR.

---

```diff
diff --git a/src/opm/output/eclipse/EclipseIO.cpp b/src/opm/output/eclipse/EclipseIO.cpp
index 93f7ed1f0..5692365d3 100644
--- a/src/opm/output/eclipse/EclipseIO.cpp
+++ b/src/opm/output/eclipse/EclipseIO.cpp
@@ -227,6 +227,15 @@ void EclipseIO::writeTimeStep(const Action::State& action_state,
       but there is an unsupported option to the RPTSCHED keyword which
       will request restart output from every timestep.
     */
+    {
+        std::stringstream ss;
+        ss << "!isSubstep && schedule.restart().getWriteRestartFile("
+           << report_step << ") = " << !isSubstep << " && "
+           << schedule.restart().getWriteRestartFile(report_step) << " = "
+           << (!isSubstep && schedule.restart().getWriteRestartFile(report_step));
+
+        OpmLog::note(ss.str());
+    }
     if(!isSubstep && schedule.restart().getWriteRestartFile(report_step))
     {
         EclIO::OutputStream::Restart rstFile {
```

**Comment by joakim-hove** [comment]
> For what it's worth, I applied the current patch and

Thank you for looking into this; I have done exactly the same as you have - and reached the same conclusion, will continue tomorrow.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this update_data please

**Comment by joakim-hove** [comment]
jenkins build this opm-tests=350 please

File: tests/parser/RestartConfigTests.cpp [hunk 1]
```diff
@@ -37,6 +37,152 @@ inline std::string fst( const std::pair< std::string, int >& p ) {
...
+
+    const auto deck   = Parser{}.parseString(input);
+    const auto tm     = TimeMap { deck };
+    const auto rstcfg = RestartConfig { tm, deck };
+
+    for (std::size_t step = 0; step < tm.size(); step++) {
+        if (step == 0 || step == 38 || step == 51)
+            BOOST_CHECK( rstcfg.getWriteRestartFile(step) );
+        else
+            BOOST_CHECK( !rstcfg.getWriteRestartFile(step) );
```
Comment (bska) [inline tests/parser/RestartConfigTests.cpp:182]: Do  you think you could use `BOOST_CHECK_MESSAGE` here and include the `step` in the message?  That will assist debugging if it ever breaks.

Comment (joakim-hove) [reply tests/parser/RestartConfigTests.cpp]: Done

**Pull request merged**